### DISCUSSION TITLE
Why this event does'nt work?

### DISCUSSION THREAD

**Discussion by shadow3aaa** [tauri-apps/tauri#10033] [Q&A] [answered]
I'm writing an rust ide on android.

Since tauri(2.0-beta) doesn't yet support mobile to use sidecar/resources, I use `rust_embed crate` to embed resources directly into the library.

Then I need to extract some resources(like proot, rootfs, prebuilted rust-analyzer, etc.) when the program starts and put them in a place like `/data/data/package_name/res`.

I want users to see the progress of resource extraction, So I do this(vue3):

- frontend
```typescript
import { ref, onMounted, onUnmounted } from "vue";
import { useRouter } from "vue-router";
import { VList, VListItem, VIcon } from "vuetify/components";
import { listen, UnlistenFn } from "@tauri-apps/api/event";
import { invoke } from "@tauri-apps/api/core";

interface Payload {
  message: string;
}

let unlisten: UnlistenFn | null = null;

const router = useRouter();
const setupProcess = ref("Waiting until setup process complete...");
const setupCompleted = ref(false);

onMounted(async () => {
  invoke("init_resources").then(() => {
    setupCompleted.value = true;
  });

  unlisten = await listen<Payload>("setup-process", (event) => {
    const { message } = event.payload;
    setupProcess.value += `\n${message}`;
  });
});

onUnmounted(() => {
  unlisten?.();
});

function routeToProjectView() {
  router.push("/project");
}
```

- backend
`payload.rs`
```rust
#[derive(Clone, serde::Serialize)]
pub struct Payload {
    message: String,
}

impl Payload {
    pub fn new(message: String) -> Self {
        Self { message }
    }
}
```

`lib.rs`
```rust
#[tauri::command]
fn init_resources(app: AppHandle) {
    app.emit("setup-process", Payload::new("Checking resources".into()))
        .unwrap();
    
    async_runtime::block_on(Resources::auto_update()).unwrap();
    app.emit(
        "setup-process",
        Payload::new("Checking proot rootfs".into()),
    )
    .unwrap();
    setup_rootfs().unwrap();
    app.emit("setup-process", Payload::new("All done.".into()))
        .unwrap();
}

#[cfg_attr(mobile, tauri::mobile_entry_point)]
pub fn run() {
    tauri::Builder::default()
        .plugin(tauri_plugin_shell::init())
        .invoke_handler(tauri::generate_handler![init_resources])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
```

However, when I was testing it, I found that no events were received, and if I tried to await listen<Payload>, it just kept blocking current thread.

**Comment by FabianLars** [comment]
Did you try changing this
```ts
onMounted(async () => {
  invoke("init_resources").then(() => {
    setupCompleted.value = true;
  });

  unlisten = await listen<Payload>("setup-process", (event) => {
    const { message } = event.payload;
    setupProcess.value += `\n${message}`;
  });
});
```
to this?
```ts
onMounted(async () => {
  unlisten = await listen<Payload>("setup-process", (event) => {
    const { message } = event.payload;
    setupProcess.value += `\n${message}`;
  });

  invoke("init_resources").then(() => {
    setupCompleted.value = true;
  });
});
```

If the command executes too fast it will be finished before you have a chance to register the events, so just make sure the listeners are registered beforehand and it should work.

**Reply by shadow3aaa** [reply]
I tried this:
```typescript
const init_resources = async () => {
  await invoke("init_resources").then(() => {
    setupCompleted.value = true;
  });
};

onMounted(async () => {
  unlisten = listen<Payload>("setup-process", (event) => {
    const { message } = event.payload;
    setupProcess.value = `\n${message}`;
  });

  init_resources();
});
```
but still, it didn't work(invoke works fine)
![图片](https://github.com/tauri-apps/tauri/assets/74781933/8af70ad5-302f-471d-8e04-4d4b85c4d010)

**Reply by FabianLars** [reply]
Can you check the devtools console (right click -> inspect) for any errors? Maybe also try to console.log the event payload in the listener to be sure it never triggers

**Reply by shadow3aaa** [reply, marked as answer]
> Can you check the devtools console (right click -> inspect) for any errors? Maybe also try to console.log the event payload in the listener to be sure it never triggers

![图片](https://github.com/tauri-apps/tauri/assets/74781933/f7694286-17ad-4f3f-a2c0-bb2b19f8d0e9)
SO THAT'S IT. There is a permission error. That's strange, I think it should be allowed by default

**Reply by FabianLars** [reply]
Well, it is if you add the _default_ permissions (`event:default`).

**Comment by shadow3aaa** [comment]
So this is my solution:

- create dir `src-tauri/capabilities`(if not created yet)
- create file `src-tauri/capabilities/mobile.json`(or `desktop.json`, `main.json`, etc):
```json5
{
  "$schema": "../gen/schemas/mobile-schema.json",
  "identifier": "main-capability",
  "description": "Capability for the main window",
  "windows": ["main"],
  "permissions": ["event:default"] // add any other permissions you need!
}
```

and event should work fine
